### PR TITLE
initial docs on async method builders

### LINKED ISSUE #24950: Allow AsyncMethodBuilder() attribute on methods
This is an advanced scenario that enables library authors to optimize away allocations on async method. 

See https://github.com/dotnet/csharplang/issues/1407

### PR DESCRIPTION
**Pull request by BillWagner**
Add and update the description of AsyncMethodBuilder, and the new feature that it can be applied to a method.

Fixes #24950 

See issue #24972 to track the remaining task.

Notes on changes are on the issue: https://github.com/dotnet/docs/issues/24950#issuecomment-872541399

### CODE REVIEW

**Review by tdykstra — APPROVED** [review]
LGTM, one question -- I didn't see a high-level intro discussion that would explain under what circumstances / in what scenarios you need to use the builder attribute rather than just use a Task<T> return type.

File: docs/csharp/language-reference/attributes/general.md [hunk 1]
```diff
@@ -19,33 +19,33 @@ If the `TRACE_ON` identifier isn't defined, the trace output isn't displayed. Ex
 
 The `Conditional` attribute is often used with the `DEBUG` identifier to enable trace and logging features for debug builds but not in release builds, as shown in the following example:
 
-:::code language="csharp" source="snippets/ConditionalExamples.cs" id="SnippetConditional" :::
```
[the remark this answers is not in the document]
Comment (BillWagner) [reply docs/csharp/language-reference/attributes/general.md:22]: @tdykstra Acrolinx marks down "id" in favor of "ID". Enough samples, and it has a measurable effect on the score.

Should I ping @MonicaRush on that one? (I'd honestly thought the guidance had changed when it kept getting flagged)

**Comment by BillWagner** [comment]
@tdykstra 

>  I didn't see a high-level intro discussion that would explain under what circumstances / in what scenarios you need to use the builder attribute rather than just use a Task return type.

I'll address that as part of #24972

File: docs/csharp/language-reference/attributes/general.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (tdykstra) [reply docs/csharp/language-reference/attributes/general.md:22]: Yes, this  should count as a false positive, Acrolinx shouldn't be applying general text rules to triple-colon statements.

**Pull request merged**